genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -0,0 +1,3394 @@
+            nn = n.dot(n)
+            if nn == 0:
+                flag = 1
+            elif nv != 0 and nn > self.FLOAT_MIN:
+                point = n * (nv / nn)
+                flag = 0
+
+        if flag == -1:
+            # Last fallback
+
+            # Normal = (v1 - v3) x (v2 - v3)
+            n = d31.cross(d32)
+            nv = n.dot(v3)
+            nn = n.dot(n)
+            point = n * (nv / nn)
+            flag = 0
+
+        return point, flag
+
+    @ti.func
+    def func_project_origin_to_line(self, v1, v2):
+        """
+        Project the origin onto the line defined by the simplex vertices.
+
+        P = v2 - ((v1 * diff) / (diff * diff)) * diff
+        """
+        diff = v2 - v1
+        k = v2.dot(diff) / diff.dot(diff)
+        P = v2 - k * diff
+
+        return P
+
+    @ti.func
+    def func_simplex_vertex_linear_comb(self, i_b, i_v, i_s1, i_s2, i_s3, i_s4, _lambda, n):
+        """
+        Compute the linear combination of the simplex vertices
+
+        @ i_v: Which vertex to use (0: obj1, 1: obj2, 2: minkowski)
+        @ n: Number of vertices to combine, combine the first n vertices
+        """
+        res = gs.ti_vec3(0, 0, 0)
+
+        s1 = self.gjk_simplex_vertex[i_b, i_s1].obj1
+        s2 = self.gjk_simplex_vertex[i_b, i_s2].obj1
+        s3 = self.gjk_simplex_vertex[i_b, i_s3].obj1
+        s4 = self.gjk_simplex_vertex[i_b, i_s4].obj1
+        if i_v == 1:
+            s1 = self.gjk_simplex_vertex[i_b, i_s1].obj2
+            s2 = self.gjk_simplex_vertex[i_b, i_s2].obj2
+            s3 = self.gjk_simplex_vertex[i_b, i_s3].obj2
+            s4 = self.gjk_simplex_vertex[i_b, i_s4].obj2
+        elif i_v == 2:
+            s1 = self.gjk_simplex_vertex[i_b, i_s1].mink
+            s2 = self.gjk_simplex_vertex[i_b, i_s2].mink
+            s3 = self.gjk_simplex_vertex[i_b, i_s3].mink
+            s4 = self.gjk_simplex_vertex[i_b, i_s4].mink
+
+        c1 = _lambda[0]
+        c2 = _lambda[1]
+        c3 = _lambda[2]
+        c4 = _lambda[3]
+
+        if n == 1:
+            res = s1 * c1
+        elif n == 2:
+            res = s1 * c1 + s2 * c2
+        elif n == 3:
+            res = s1 * c1 + s2 * c2 + s3 * c3
+        else:
+            res = s1 * c1 + s2 * c2 + s3 * c3 + s4 * c4
+        return res
+
+    @ti.func
+    def func_det3(self, v1, v2, v3):
+        """
+        Compute the determinant of a 3x3 matrix formed by the vectors v1, v2, v3.
+        M = [v1 | v2 | v3]
+        """
+        return (
+            v1[0] * (v2[1] * v3[2] - v2[2] * v3[1])
+            - v1[1] * (v2[0] * v3[2] - v2[2] * v3[0])
+            + v1[2] * (v2[0] * v3[1] - v2[1] * v3[0])
+        )
+
+    @ti.func
+    def func_compare_sign(self, a, b):
+        """
+        Compare the sign of two values.
+        """
+        ret = 0
+        if a > 0 and b > 0:
+            ret = 1
+        elif a < 0 and b < 0:
+            ret = -1
+        return ret
+
+    @ti.func
+    def func_is_equal_vec3(self, a, b):
```

Comment:
What about making this a standalone helper function? It has nothing to do with `GJK` actually.

By the way, the name is not good, it should rather be `func_is_equal_vec`

Reply:
You can either add it on top of this file, or create a new file `genesis/utils/misc.py`.